[RDY][CR] Extend the inventory UI framework and customize gunmod installation menu

codemime opened the pull request:
Continues #17765.
Closes #17074 (milestone).
Closes #17427.
Fixes #18661.
### Observable behavior

The new customized menu shows annotations, installation odds (no repair kit modifiers yet) and disables all invalid choices (they're grayed out and cannot be selected). Note that no unrelated information is shown:
![01](https://cloud.githubusercontent.com/assets/11905613/18915272/7e8e1204-85a1-11e6-8886-47ec72042d55.png)

 All entries are sorted using this priority (in descending order): 
- has greater odds
- enabled/disabled
- alphabetically.

Unused columns (**damage** in this case) are omitted. 100% of successful installation is marked as "always":
![02](https://cloud.githubusercontent.com/assets/11905613/18915297/93cf62d0-85a1-11e6-8fc3-1102d6bfaf41.png)

When all entries are disabled, it looks like this:
![03](https://cloud.githubusercontent.com/assets/11905613/18915359/c9ae0aaa-85a1-11e6-8c9b-bf4f193d3c1a.png)

Inventory screen. Note the new header:
![screenshot_2016-09-24_22-11-0](https://cloud.githubusercontent.com/assets/11905613/18810683/7a397d8e-82ad-11e6-8a24-1acb6cd36c7c.png)

Weight/volume are insignificant for the use menu and therefore omitted.
![screenshot_2016-09-24_22-08-07](https://cloud.githubusercontent.com/assets/11905613/18810680/7a33d17c-82ad-11e6-8729-d6d95a620608.png)

HDMI resolution. Note that **success** and **damage** are shown only in the left column. Columns are left-aligned because they occupy less than 65% of the screen:
![04](https://cloud.githubusercontent.com/assets/11905613/18915366/cf874d88-85a1-11e6-8773-490c3855f16e.png)
### Implementation details
- Presets now have a `get_denial()` function which returns the reason why this item (location) cannot be selected. When the function returns a non-empty string, the entry gets disabled (grayed out)
- Entry sorting using `sort_compare()`. Enabled items always go before disabled ones and by-default sorted alphabetically
- Smarter cell representation. Empty cells (those that don't contain any text for the shown items) become hidden along with their headers. Spaces between headers were increased
- Introduced `gunmod_inventory_preset` (preset for gunmods)
- Gunmod installation odds get calculated in a separate function
- Drawing is now split into separate functions (`draw_header()`, `draw_footer()`, `draw_columns()`). No need for virtual `inventory_selector::draw()` anymore
- Reduced code duplication in `game_inventory.cpp`
- Reduced (eliminated?) costs of verifying item location. There's no need until we actually want to do something with the item (inventory screen is not the case)
- Fixed a bug where gunmod installation roll could become negative
- More changes. See the code for details.

~~WIP because I want to test it some more and comment a bit better.~~

mugling commented:
How practical would it be to split the interface vertically and then show the effects of adding the gunmod? I appreciate this is a little off your original plan but how practical could that be?

codemime commented:
> I appreciate this is a little off your original plan but how practical could that be?

It's not off-the-plan at all: the idea occurred to me too, and I quite like it. The current design with flexible columns and `preset`s allows to implement it fairly easy. I meant if for general interface screen in the first place (inspecting each item separately is too tiresome), but later: one major change at a time.

codemime commented:
Ready to be reviewed/tested (according to #18563, the merge conflict is considered temporary).

mugling commented on src/game.h:
```diff
@@ -472,8 +471,7 @@ class game
                                       const std::string &none_message = "" );
-        item_location inv_map_splice( item_location_filter filter, const std::string &title, int radius = 0,
-                                      const std::string &none_message = "" );
+        item_location inv_for_gunmod( const item &gunmod, const std::string &title );
```
Functions should preferably have doxygen comments

codemime replied:
Added.

mugling commented on src/game_inventory.cpp:
```diff
@@ -18,6 +18,31 @@ class inventory_filter_preset : public inventory_selector_preset
+    inv_s.add_nearby_items( radius );
+
+    if( inv_s.empty() ) {
+        const std::string msg = ( none_message.empty() ) ? _( "You don't have the necessary item at hand." )
```
Redundant parenthesis. Also could probably fit one one line with `auto`

codemime replied:
Fixed, but still doesn't fit into one line.

mugling commented on src/game_inventory.cpp:
```diff
@@ -154,24 +154,100 @@ item *game::inv_map_for_liquid( const item &liquid, const std::string &title, in
+                                        liquid.tname().c_str() ) ).get_item();
+}
+
+class gunmod_inventory_preset: public inventory_selector_preset
```
space after class name

mugling commented on src/game_inventory.cpp:
```diff
@@ -154,24 +154,100 @@ item *game::inv_map_for_liquid( const item &liquid, const std::string &title, in
+{
+    public:
+        gunmod_inventory_preset( const player &p, const item &gunmod ) : p( p ), gunmod( gunmod ) {
+            append_cell( [ this ]( const item_location & loc ) -> std::string {
```
avoid specifying return type for lambda if necessary. consistency with existing code is preferable. subjective because if you have a prevailing technical argument that precedent could be changed

codemime replied:
Fixed.

mugling commented on src/game_inventory.cpp:
```diff
@@ -154,24 +154,100 @@ item *game::inv_map_for_liquid( const item &liquid, const std::string &title, in
+
+            const auto &min_skills = gunmod.type->min_skills;
+            const auto &gun_skill = loc->gun_skill();
+            const auto iter = min_skills.find( "weapon" );
```
Can you document what is going on here a little more

codemime replied:
Added.

mugling commented on src/game_inventory.cpp:
```diff
@@ -154,24 +154,100 @@ item *game::inv_map_for_liquid( const item &liquid, const std::string &title, in
+        }
+
+    protected:
+        std::pair<int, int> get_odds( const item_location &gun ) const {
```
Needs doxygen comments explaining the pair members

codemime replied:
Added.

mugling commented on src/game_inventory.cpp:
```diff
@@ -154,24 +154,100 @@ item *game::inv_map_for_liquid( const item &liquid, const std::string &title, in
 
-    inventory_filter_preset preset(
-    [ this ]( const item_location & location ) -> bool {
+    const inventory_filter_preset preset( [ this ]( const item_location & location ) {
```
astyle regression

codemime replied:
What do you mean? I don't see any, except the space after `&` which is a glitch of astyle.

mugling replied:
Ok, disregard

mugling commented on src/inventory_ui.cpp:
```diff
@@ -160,28 +182,68 @@ void inventory_selector_preset::append_cell( const std::function<std::string( co
+    const auto get_incremented = [ this ]( size_t index, int step ) -> size_t {
+        return ( index + step + entries.size() ) % entries.size();
+    };
+
```
Can you comment this algorithm to aid future maintainers

codemime replied:
Added.

mugling commented on src/inventory_ui.cpp:
```diff
@@ -734,31 +836,79 @@ void inventory_selector::prepare_layout()
+    static const int cells_count = 4;
+
+    using stat = std::array<std::string, cells_count>;
+    const auto disp = []( const std::string &caption, int cur_value, int max_value,
```
Can you document this lambda explaining what it does. Doesn't need to be expansive.

codemime replied:
Added.

mugling replied:
Where did you add this?

mugling replied:
Did you rebase as opposed to fix merge conflicts?

mugling commented on src/inventory_ui.cpp:
```diff
@@ -734,31 +836,79 @@ void inventory_selector::prepare_layout()
+        disp( string_format( _( "Volume (%s):" ), _( "L" ) ),
+              units::to_milliliter( dummy.volume_carried() ),
+              units::to_milliliter( dummy.volume_capacity() ), []( int v ) {
+            return string_format( "%.1f", round_up( double( v ) / 1000, 1 ) );
```
What happened to `%6.1f`?

codemime replied:
I don't need fixed width anymore: it became adjustable.

mugling commented on src/item.cpp:
```diff
@@ -4221,60 +4221,62 @@ const item * item::gunmod_find( const itype_id& mod ) const
     return const_cast<item *>( this )->gunmod_find( mod );
 }
 
-bool item::gunmod_compatible( const item& mod, bool alert, bool effects ) const
```
What are the consequences of dropping the `effects` parameter?

codemime replied:
None. When one installs a gunmod in ...

the current master:

- `gunmod_compatible( effects = false )` gets called for each gun to filter compatible ones
- `gunmod_compatible( effects = true )` gets called for selected gun to reject it if it's loaded (fiasco).

this PR:

- `gunmod_compatible()` gets called for each gun only once and filters out incompatible or loaded ones. It shows an appropriate denial message in either case.

mugling commented on src/player.cpp:
```diff
@@ -11107,9 +11107,43 @@ bool player::gunmod_remove( item &gun, item& mod )
     return true;
 }
 
+/** @return Odds for success (pair.first) and gunmod damage (pair.second) */
```
Doxygen comments in the header, not here

codemime replied:
Fixed.

codemime commented:
> Design is good. Fairly clean code.

Thank you.

mugling commented on src/game.h:
```diff
@@ -471,6 +471,10 @@ class game
         item_location inv_map_splice( item_filter filter, const std::string &title, int radius = 0,
                                       const std::string &none_message = "" );
+        /**
+          * Execute customized inventory menu for the specified @ref gunmod.
```
Presuming there is going to be more of these you could add doxygen family documentation:
```
/**
 * @name Item selection interface
 *
 * Your documentation here describing how the functions work
 */
/*@{*/

/** Very brief description only, fetches gunmods... */
item_location inv_for_gunmod( ...)


/** Very brief description only, fetches foo.. */
item_location inv_for_foo( ...)

/*@}*/
```

mugling replied:
Good examples of this are found in `item.h`

codemime replied:
Noted.

codemime commented:
All set.

codemime commented:
@mugling I'd appreciate if you could test/merge this PR before long. I have a couple more PRs based on this one.

codemime commented:
Rebased.

mugling commented:
Doesn't work. All mods display 'beyond your skill'

codemime commented:
Oops! An accidental corruption during the rebase. The fix is trivial: see https://github.com/CleverRaven/Cataclysm-DDA/pull/18514/commits/598d44e46b4cd78fc16cacced617ee96ec5d4e8b

codemime commented:
> Doesn't work. All mods display 'beyond your skill'

It does work now. When will you give it another try?

mugling commented:
A good idea, but:

* Item names and denial messages are often concatenated making it difficult to read through the list of candidates. Needs to be a minimum spacing after item name or similiar plus possibly better use of coloring (perhaps names should be white or possibly green/red dependent upon status)

* Column spacing is problematic. Sometimes 20% of screen width is used, other times 100%. Seems to depend somewhat on length of denial messages

* The most common task is to add a gunmod to the wielded item. The `RIGHT` selection issue makes this task unintuitive

* Conversion kits have bad wording 'M1911 with 5.7mm kit doesn't have slot for this mod`

mugling commented:
Not introduced by your PR at all but #18661 could be fixed here (as any fix is otherwise a potential merge conflict)

mugling commented:
Also confusing what do `success` and `damage` mean? One option is adding a 'description' pane to this UI that could show short help text etc?

codemime commented:
> Item names and denial messages are often concatenated making it difficult to read through the list of candidates.

That was the initial idea: complete sentences, but spaces can be added easily (there's a constant for that).

> Conversion kits have bad wording 'M1911 with 5.7mm kit doesn't have slot for this mod`

The very same message would be printed if you try to install this mod without this PR.

> Column spacing is problematic. Sometimes 20% of screen width is used, other times 100%. Seems to depend somewhat on length of denial messages

It depends on occupation ratio. If columns occupy more than 65% of the screen, they occupy all 100% (can be tweaked easily - there's a constant for that).

> The most common task is to add a gunmod to the wielded item. The RIGHT selection issue makes this task unintuitive

I'll fix it as soon as a good usage scenario appears.

> Not introduced by your PR at all but #18661 could be fixed here (as any fix is otherwise a potential merge conflict)

No problem.

> Also confusing what do success and damage mean? One option is adding a 'description' pane to this UI that could show short help text etc?

Description pane will follow in later PRs. Now I need short column headers. Any suggestions of what to use instead of `success` and `damage`?

mugling commented:
> That was the initial idea: complete sentences, but spaces can be added easily (there's a constant for that).

Complete sentences aren't necessarily problematic but would be clearer with highlighting, spaces or possibly both.

> The very same message would be printed if you try to install this mod without this PR.

Can you fix it here?

> It depends on occupation ratio. If columns occupy more than 65% of the screen, they occupy all 100% (can be tweaked easily - there's a constant for that).

Why not 100% all the time? Or alternatively an algorithm that more gradually adjusts the layout?

> I'll fix it (ed: `RIGHT`) as soon as a good usage scenario appears.

That display could work better as one column thereby avoiding this issue. Would also allow for display of more statistics columns

> Description pane will follow in later PRs. Now I need short column headers. Any suggestions of what to use instead of success and damage?

I don't have a suggestion for simpler names but it's arguably a regression from the previous (granted already imperfect) implementation. Consider that I wrote that code and I'm not certain what you mean. You probably need to tackle the repair kit percentages as part of this PR.

mugling commented on src/game_inventory.cpp:
```diff
@@ -172,12 +172,12 @@ class gunmod_inventory_preset : public inventory_selector_preset
                 const auto odds = get_odds( loc );
                 return odds.second > 0 ? string_format( "<color_red>%d%%</color>", odds.second ) : std::string();
-            }, _( "DAMAGE" ) );
+            }, _( "RISK OF DAMAGE" ) );
```
`DAMAGE RISK`?

mugling commented on src/game_inventory.cpp:
```diff
@@ -172,12 +172,12 @@ class gunmod_inventory_preset : public inventory_selector_preset
 
                 return string_format( "<color_ltgreen>%d%%</color>", odds.first );
-            }, _( "SUCCESS" ) );
+            }, _( "ODDS OF SUCCESS" ) );
```
`SUCCESS CHANCE`?

codemime commented:
> Complete sentences aren't necessarily problematic but would be clearer with highlighting, spaces or possibly both.

I'm against highlighting: they are disabled and, hence, should be grayed out. Spaces also don't look, why would we have intentional spaces in complete sentences?

> Can you fix it here?

If you suggest a good way of fixing it. Otherwise, no: it's a pre-existed and rather minor thing.

> Why not 100% all the time? Or alternatively an algorithm that more gradually adjusts the layout?

Because it looks too damn ugly: narrow columns in different parts of the screen. The current layout looks good enough.

> That display could work better as one column thereby avoiding this issue. Would also allow for display of more statistics columns

One column would be a regression. Again, I'm going to implement an information panel (as I've said before), but not here. The PR is big enough already. Also, it becomes one column if you bother to test it on smaller resolutions (I guess a proper testing Implies this).

mugling commented:
> One column would be a regression.

Why? How does the display of typically one or two items in multiple columns aid usability?

codemime commented:
> Why? How does the display of typically one or two items in multiple columns aid usability?

Because it's consistent with the inventory (where are lots of items and it's desirable to see them all at once without scrolling). However, I guess this can be done if columns are really small. But, again, not here, there are more following PRs.

codemime commented:
Rebased and ready.

mugling commented:
The screenshots (good) aren't whats typically displayed

codemime commented:
What do you mean, the renamed columns)?

mugling commented:
It shows everything in one column and with good spacing between the items and reasons

codemime commented:
It means that your items can't fit into one column and the "suqashed" column is wider than 65% of your screen. Can you attach a screenshot of what you consider wrong and we'll see what we can do about it?

mugling commented:
It would be best if display for gunmods was always like the screenshots

codemime commented:
> It would be best if display for gunmods was always like the screenshots

All combinations of gunmods on all resolutions? It's quite a lot of screenshots. Screenshots displaying wrong layouts are far more useful. What's yours?

mugling commented:
Well your screenshots show the desired output - single column with spacing between item and reason

codemime commented:
See the last one: there are two. Can you just upload your screenshot here with what you consider undesirable output?

mugling commented:
Good point, I didn't scroll down - I'll try for some screenshots

codemime commented:
I have another PR which improves space management and customizes the reading menu. It's almost ready, but I can't push it, as it's based on this one and already needs rebasing. Post your screenshots whenever you like, make your suggestions: I'll take them into account in the next PR, but please, don't detain development. I guess there's no flaws that prevent merging.

mugling commented:
> Post your screenshots whenever you like

I'm trying to compile an image editor now to annotate some screenshots but the below should hopefully be clear

> I have another PR which improves space management and customizes the reading menu

Spacing between columns needs to be improved - at the moment they either use too little of the screen or too much. It would be reasonable to defer this to your next PR

> make your suggestions:

There should be a minimum space between the gun name and the denial reason. This could do with fixing for this PR even if it just involves adding 3 or 4 spaces between the two.

Also the `RIGHT` selection bug needs fixing as we have players activating grenades in error and the like. A more robust/unified approach can come later but for now having `LEFT` and `RIGHT` be opposites (but inconsistent with multidrop) is much preferable to having `RIGHT` be a game-ending mistake.

> I guess there's no flaws that prevent merging.

No, I couldn't find any more bugs other than the resolved issues with the skill check. The `RIGHT` fix is presumably trivial and I don't suppose fixing spacing after the item names would be difficult. If you don't want to fix column spacing that can be deferred to your next PR

mugling commented:
+1 for the YASD reference

codemime commented:
Okay, changed behavior of the arrow keys. No more YASDs. Also, small code improvements and added handy hints for empty selections. See the last commits.

`RETURN` was pressed:
![x](https://cloud.githubusercontent.com/assets/11905613/19224896/1fa1d8ea-8ea0-11e6-8881-0e9d1173432e.png)

codemime commented:
Ready. Other changes will be discussed in the next PR.

mugling commented:
Did you add spacing between gun name and reason?

codemime commented:
No, I think columns should be as narrow as possible. Also, huge spaces in complete sentences aren't desirable.

mugling commented:
Your first screenshot is much more readable than the final (discounting the changes in screen size)

codemime commented:
> Your first screenshot is much more readable than the final (discounting the changes in screen size)

I admit it's not perfect yet, but readability is an arguable matter in this case (especially keeping in mind that in
the current master denials for gunmods aren't readable at all). Please report if you find a bug, I shall fix it here, but other changes can be discussed in the next PR. I consider this one finished as it's already a big improvement compared to the current gunmods selection.

mugling commented:
> That was the initial idea: complete sentences, but spaces can be added easily (there's a constant for that).

Can you bump that constant for now as this PR is otherwise good to go?

codemime commented:
> Can you bump that constant for now as this PR is otherwise good to go?

No, the constant is in the next PR (along with a few more):

```c++
/** The minimal gap between two cells */
static const int min_cell_gap = 2;
/** The gap between two cells when screen space is limited*/
static const int normal_cell_gap = 4;
/** The minimal gap between the first cell and denial */
static const int min_denial_gap = 1;
/** The minimal gap between two columns */
static const int min_column_gap = 2;
/** The gap between two columns when there's enough space, but they are not centered */
static const int normal_column_gap = 8;
/** The gap on the edges of the screen */
static const int screen_border_gap = 1;
/** The minimal occupancy ratio (see @refer get_columns_occupancy_ratio()) to align columns to the center */
static const double min_ratio_to_center = 0.65;
```
As I said, I won't add any new stuff here because I don't want make further rebasing any harder than it should be for no reason.

mugling commented:
You could push at least `min_denial_gap` as part of this PR as it's really required here to prevent a regression. Also push your next proposed PR (dependent upon this) as the only real obstacle to merging is now that spacing

codemime commented:
> the only real obstacle to merging is now that spacing
> as it's really required here to prevent a regression

That's ridiculous. How can it possibly be a "real obstacle"? There's nothing in the current master to regress until there's something. I don't agree with this change, I won't add it here and I won't argue anymore.

mugling commented:
Your screenshots don't look like the output of your PR? It's not unreasonable to ask that the output is updated to match. You said there was a constant but it doesn't appear to be part of this PR?

It's less readable with no boundary between the item name and the reason. It concerns me this is just going to get merged and forgotten about. You only need to add that constant to be able to mere this PR.

Think about it - the output doesn't match what your screenshots propose and you're referring to code that doesn't (as far as this repo is concerned) exist yet. I've given you two approaches to fixing that.

codemime commented:
> Think about it - the output doesn't match what your screenshots propose 

Well, I thought about it and I don't see any discrepancy, but I see that you were asked **four** times to show any screenshot that shows it, but you didn't. 

> It's less readable with no boundary between the item name and the reason.

As I said, I don't agree with that. You can make a PR that bumps the constant if you like.

> and you're referring to code that doesn't (as far as this repo is concerned) exist yet. I've given you two approaches to fixing that.

[Here's the code I refer](https://github.com/codemime/Cataclysm-DDA/tree/ref_inv_n6). I won't PR an unfinished branch.

mugling commented:
> Well, I thought about it and I don't see any discrepancy, but I see that you were asked four times to show any screenshot that shows it, but you didn't.

There is a gap between the item name and the denial reason in the first screenshots which is missing in the last one.

> Here's the code I refer. I won't PR an unfinished branch.

It looks fairly trivial to add `min_denial_gap` to this PR and set it to 4?

codemime commented:
> There is a gap between the item name and the denial reason in the first screenshots which is missing in the last one.

Yes, the screenshots _perfectly_ match the PR. The last one shows two columns (aligned to the left) and the rest are center-aligned (that's why the gap appears).

>  It looks fairly trivial to add min_denial_gap to this PR and set it to 4?

Yes, it looks like, but the constant is a part of the next PR, and I don't want it to be 4. You can make a PR that bumps the constant if you like.

mugling commented:
> Yes, it looks like, but the constant is a part of the next PR, and I don't want it to be 4. You can make a PR that bumps the constant if you like.

Now I think we get to the crux of the matter. The first screenshots are good but the last is poorly readable and unfortunately that pathological case is also the most common screen that will be displayed.

I appreciate that refactoring bring it's own advantages but overall the usability of the inventory menus has regressed. I understand that's going to bring some considerable net benefits but there are limits to how much regression is tolerable and adding spacing is an easy way of controlling that for now.

I don't want to merge this and then start debating the spacing given it's an easy change to make now.

illi-kun commented:
> I understand that's going to bring some considerable net benefits but there are limits to how much regression is tolerable

@mugling could you show some screenshots (at least the worst one), to show that major regression and demonstrate it makes merging of PR (which fixes 3 issues) impossible?

mugling commented:
This is what you get in almost all cases:
![a](https://cloud.githubusercontent.com/assets/12238606/19251826/3cebf7d4-8f39-11e6-8596-6606b6f92e9e.png)

Which looks nothing like the screenshot at the start of this PR, which you never get:
![b](https://cloud.githubusercontent.com/assets/12238606/19251849/64d5dd28-8f39-11e6-9687-9b24759f5543.png)

So three issues:
* Spacing (yellow lines) can be resolved in a later PR
* Spacing (blue lines) is easily solved in this PR by a one line change
* Screenshots posted at start of PR aren't at all representative of the output of this PR

I'm going to factor out the auxiliary gunmod fix and merge that now as it's a single line change. The `RIGHT` cursor change is welcome but belongs in this PR.

illi-kun commented:
Thank you for pics, "a picture is worth a thousand words".
- Issue with 95% emptiness on huge screen sizes is something we have for years but I agree it's better to see this fixed.
- Issue with lack of space between name and description is inconvenient but I surprised you call it "major regression", especially with the fact @codemime have some solution in work.

It's good to know that you'll not block this PR completely and ready to merge it. I hope that your changes will not discourage @codemime from further progress on this.

mugling commented:
> It's good to know that you'll not block this PR completely and ready to merge it

No, it's good to go apart from that change. Whilst the screenshots in the preamble were no doubt well intentioned they have proven problematic here in that we have only detected at this late stage that they are not representative - I would have objected to horizontal paging of continuous prose immediately.

> Yes, it looks like, but the constant is a part of the next PR, and I don't want it to be 4. You can make a PR that bumps the constant if you like.

I'm requesting a trivial change here to make the output look a bit more like the screenshots. What is going to be different when I open a PR versus just adding it now?

illi-kun commented:
> No, it's good to go apart from that change.

Yes, that's what I've tried to said, sorry for my poor language.


> I'm requesting a trivial change here to make the output look a bit more like the screenshots. What is going to be different when I open a PR versus just adding it now?

I see some contradiction with your previous quote "merge it now, fix it later".

mugling commented:
> > I'm requesting a trivial change here to make the output look a bit more like the screenshots. What is going to be different when I open a PR versus just adding it now?

> I see some contradiction with your previous quote "merge it now, fix it later".

Sorry, you lost me on that one. The first quote is from @codemime earlier.

illi-kun commented:
@mugling sorry for excessive pressure, I just don't want to see this PR stalled.

mugling commented:
> @mugling sorry for excessive pressure, I just don't want to see this PR stalled.

No me neither, so all of the issues that are difficult to resolve can be deferred leaving only the spacing change between item and reason. Overall the output does need to be adjusted to be more similar to the majority of the screenshots but most of that can come in subsequent PR's.

Coolthulhu commented:
The HD version looks too squished. There is no clear distinction between gun name and status description.
Would it be possible to let it spread into free space?

Alternatively, status could be red/green to contrast with gun name being white/gray.

Coolthulhu commented:
A more minor issue is that on wide screens, the weight/volume display "escapes" far to the right.
Fixing this could be hard since it would require communication between the main display and contents, so I guess it's tolerable.

Would be best if weight/volume was above the rightmost column displayed.

Alternatively, the whole display could be centered, though that would probably require some serious redesign.

codemime commented:
> A more minor issue is that on wide screens, the weight/volume display "escapes" far to the right.
Fixing this could be hard since it would require communication between the main display and contents, so I guess it's tolerable. Would be best if weight/volume was above the rightmost column displayed.

I thought about that too, but have rejected: if there's only one narrow column, this layout leaves no space in the header.

> Alternatively, the whole display could be centered, though that would probably require some serious redesign.

That would be fairly easy, no redesign required. I've got an alternative suggestion though. The borderless "fullscreen" window is shown whenever free space is in short supply. Otherwise, we simply don't need to fight against it:

![ready](https://cloud.githubusercontent.com/assets/11905613/19703102/a912d38e-9b13-11e6-8de5-e72a693fce9a.png)

There's another advantage: pick-up dialog can be done this way (it's quite ugly now).

codemime commented:
Rebased and ready.

> Alternatively, status could be red/green to contrast with gun name being white/gray. 

Done.

Coolthulhu commented:
Minor issues I've noticed:
* When installing a .45 conversion kit, M2 Browning had a problem description starting with "with", like "with .45 kit can't be used on M2"
* Trying to apply an incompatible gunmod by using item keybinds results in debugmsg
* Item keybinds aren't shown. They seem to work, just aren't displayed
* "x must be unloaded" message is shown before "mod can't be used on type of x" message, meaning that you have to unload things only to see that you wouldn't be able to mod them anyway

codemime commented:
Thanks for the vigilant testing. I'll fix them in a following PR.
